File: DotNet/CesiumLanguageWriterTests/TestUnitQuaternion.cs
using System;
using System.Diagnostics.CodeAnalysis;
using CesiumLanguageWriter;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    [TestFixture]
    public class TestUnitQuaternion
    {
        /// <summary>
        /// Tests that initialization of and access to the type's values works correctly.
        /// </summary>
        [Test]
        public void TestHoldValue()
        {
            UnitQuaternion quaternion = new UnitQuaternion(2.0, 4.0, 7.0, 10.0);
            Assert.AreEqual(2.0 / 13.0, quaternion.W);
            Assert.AreEqual(4.0 / 13.0, quaternion.X);
            Assert.AreEqual(7.0 / 13.0, quaternion.Y);
            Assert.AreEqual(10.0 / 13.0, quaternion.Z);
        }

        [Test]
        public void TestInitializeAndReturnMagnitude()
        {
            double magnitude;
            UnitQuaternion quaternion = new UnitQuaternion(2.0, 4.0, 7.0, 10.0, out magnitude);
            Assert.AreEqual(2.0 / 13.0, quaternion.W, Constants.Epsilon15);
            Assert.AreEqual(4.0 / 13.0, quaternion.X, Constants.Epsilon15);
            Assert.AreEqual(7.0 / 13.0, quaternion.Y, Constants.Epsilon15);
            Assert.AreEqual(10.0 / 13.0, quaternion.Z, Constants.Epsilon15);
            Assert.AreEqual(13.0, magnitude, Constants.Epsilon15);
        }

        [TestCase(60.0, 2.0, 3.0, 6.0, Description = "60 degrees, rotation about 2/7, 3/7, 6/7 vector.")]
        [TestCase(120.0, 6.0, -3.0, -2.0, Description = "120 degrees, rotation about 6/7, -3/7, -2/7 vector.")]
        [TestCase(120.0, -2.0, -3.0, 6.0, Description = "120 degrees, rotation about -2/7, -3/7, 6/7 vector.")]
        [TestCase(120.0, -2.0, 6.0, -3.0, Description = "120 degrees, rotation about -2/7, 6/7, -3/7 vector.")]
        public void TestFromMatrix3By3(double angleDegrees, double axisX, double axisY, double axisZ)
        {
            Cartesian axis = new Cartesian(axisX, axisY, axisZ);
            UnitCartesian unit = axis.Normalize();

            double angle = angleDegrees * Constants.RadiansPerDegree;
            double c = Math.Cos(angle);
            double s = Math.Sin(angle);

            double w = c;
            double x = s * unit.X;
            double y = s * unit.Y;
            double z = s * unit.Z;

            UnitQuaternion quaternion = new UnitQuaternion(w, x, y, z);
            Matrix3By3 matrix = new Matrix3By3(quaternion);

            UnitQuaternion test = new UnitQuaternion(matrix);

            Assert.AreEqual(w, quaternion.W, Constants.Epsilon15);
            Assert.AreEqual(x, quaternion.X, Constants.Epsilon15);
            Assert.AreEqual(y, quaternion.Y, Constants.Epsilon15);
            Assert.AreEqual(z, quaternion.Z, Constants.Epsilon15);
        }

        /// <summary>
        /// Tests the equality and inequality methods and operators.
        /// </summary>
        [Test]
        [SuppressMessage("Assertion", "NUnit2010", Justification = "This is specifically testing equality methods and operators")]
        public void TestEquality()
        {
            UnitQuaternion first = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
            UnitQuaternion second = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
            Assert.AreEqual(first, second);
            Assert.AreEqual(second, first);
            Assert.IsTrue(first == second);
            Assert.IsTrue(second == first);
            Assert.IsFalse(first != second);
            Assert.IsFalse(second != first);
            Assert.IsTrue(first.Equals(second));
            Assert.IsTrue(second.Equals(first));

            second = new UnitQuaternion(1.1, 2.0, 3.0, 4.0);
            Assert.AreNotEqual(first, second);
            Assert.AreNotEqual(second, first);
            Assert.IsFalse(first == second);
            Assert.IsFalse(second == first);
            Assert.IsTrue(first != second);
            Assert.IsTrue(second != first);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));

            second = new UnitQuaternion(1.0, 2.2, 3.0, 4.0);
            Assert.AreNotEqual(first, second);
            Assert.AreNotEqual(second, first);
            Assert.IsFalse(first == second);
            Assert.IsFalse(second == first);
            Assert.IsTrue(first != second);
            Assert.IsTrue(second != first);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));

            second = new UnitQuaternion(1.0, 2.0, 3.3, 4.0);
            Assert.AreNotEqual(first, second);
            Assert.AreNotEqual(second, first);
            Assert.IsFalse(first == second);
            Assert.IsFalse(second == first);
            Assert.IsTrue(first != second);
            Assert.IsTrue(second != first);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));

            second = new UnitQuaternion(1.0, 2.0, 3.3, 4.4);
            Assert.AreNotEqual(first, second);
            Assert.AreNotEqual(second, first);
            Assert.IsFalse(first == second);
            Assert.IsFalse(second == first);
            Assert.IsTrue(first != second);
            Assert.IsTrue(second != first);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));
        }

        [Test]
        public void TestEqualsEpsilon()
        {
            UnitQuaternion first = new UnitQuaternion(1.0, 1.0, 1.0, 1.0);
            UnitQuaternion second = new UnitQuaternion(0.99, 1.0, 1.0, 1.01);

            Assert.IsTrue(second.EqualsEpsilon(first, Constants.Epsilon1));
            Assert.IsTrue(second.EqualsEpsilon(first, Constants.Epsilon2));
            Assert.IsFalse(second.EqualsEpsilon(first, Constants.Epsilon3));
            Assert.IsFalse(second.EqualsEpsilon(first, Constants.Epsilon4));
            Assert.IsFalse(second.EqualsEpsilon(first, Constants.Epsilon5));
        }

        /// <summary>
        /// Tests to ensure the equality fails when comparing incorrect type.
        /// </summary>
        [Test]
        [SuppressMessage("ReSharper", "SuspiciousTypeConversion.Global")]
        [SuppressMessage("Assertion", "NUnit2010", Justification = "This is specifically testing equality methods")]
        public void TestEqualityWithWrongType()
        {
            UnitQuaternion first = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
            Cartographic second = new Cartographic(1.0, 2.0, 3.0);

            Assert.IsFalse(first.Equals(second));
        }

        /// <summary>
        /// Tests that rotation by an invalid value produces an
        /// <see cref="NotFiniteNumberException"/>.
        /// </summary>
        [Test]
        public void TestFromInfinity()
        {
            Assert.Throws<NotFiniteNumberException>(() =>
            {
                var unused = new UnitQuaternion(double.PositiveInfinity, 0.0, 0.0, 0.0);
            });
        }

        [Test]
        public void TestConjugate()
        {
            UnitQuaternion original = new UnitQuaternion(2.0, 4.0, 7.0, 10.0);
            UnitQuaternion conjugate = original.Conjugate();
            Assert.AreEqual(2.0 / 13.0, conjugate.W);
            Assert.AreEqual(-4.0 / 13.0, conjugate.X);
            Assert.AreEqual(-7.0 / 13.0, conjugate.Y);
            Assert.AreEqual(-10.0 / 13.0, conjugate.Z);
        }

        [Test]
        public void TestIdentity()
        {
            UnitQuaternion identity = UnitQuaternion.Identity;
            Assert.AreEqual(1.0, identity.W);
            Assert.AreEqual(0.0, identity.X);
            Assert.AreEqual(0.0, identity.Y);
            Assert.AreEqual(0.0, identity.Z);
        }

        [Test]
        public void TestIsUndefined()
        {
            Assert.IsFalse(new UnitQuaternion(1.0, 1.0, 1.0, 1.0).IsUndefined);
            Assert.IsTrue(UnitQuaternion.Undefined.IsUndefined);
            Assert.IsTrue(new UnitQuaternion(double.NaN, 1.0, 1.0, 1.0).IsUndefined);
            Assert.IsTrue(new UnitQuaternion(1.0, double.NaN, 1.0, 1.0).IsUndefined);
            Assert.IsTrue(new UnitQuaternion(1.0, 1.0, double.NaN, 1.0).IsUndefined);
            Assert.IsTrue(new UnitQuaternion(1.0, 1.0, 1.0, double.NaN).IsUndefined);
        }

        [Test]
        public void TestNegation()
        {
            UnitQuaternion u = -new UnitQuaternion(2.0, 4.0, 7.0, 10.0);
            Assert.AreEqual(-2.0 / 13.0, u.W);
            Assert.AreEqual(-4.0 / 13.0, u.X);
            Assert.AreEqual(-7.0 / 13.0, u.Y);
            Assert.AreEqual(-10.0 / 13.0, u.Z);
        }

        [Test]
        public void TestMultiplicationByUnitQuaternion()
        {
            // Choose quaternions whose vector portions (x, y, z) do not produce zeros
            // when dotted or crossed.  They should also have non-zero scalar (w) portions as well.
            UnitQuaternion first = new UnitQuaternion(7.0, 2.0, 3.0, 6.0);
            UnitQuaternion second = new UnitQuaternion(7.0, 2.0, 3.0, 6.0);

            UnitQuaternion result = first.Multiply(second);
            Assert.AreEqual(0.0 / 7.0, result.W, Constants.Epsilon15);
            Assert.AreEqual(2.0 / 7.0, result.X, Constants.Epsilon15);
            Assert.AreEqual(3.0 / 7.0, result.Y, Constants.Epsilon15);
            Assert.AreEqual(6.0 / 7.0, result.Z, Constants.Epsilon15);

            result = first * second;
            Assert.AreEqual(0.0 / 7.0, result.W, Constants.Epsilon15);
            Assert.AreEqual(2.0 / 7.0, result.X, Constants.Epsilon15);
            Assert.AreEqual(3.0 / 7.0, result.Y, Constants.Epsilon15);
            Assert.AreEqual(6.0 / 7.0, result.Z, Constants.Epsilon15);
        }

        /// <summary>
        /// Tests that GetHashCode returns something at least reasonably random.
        /// </summary>
        [Test]
        public void TestGetHashCode()
        {
            var object1 = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
            var object2 = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
            var object3 = new UnitQuaternion(1.0, 2.0, 3.0, 4.1);
            Assert.AreEqual(object1.GetHashCode(), object2.GetHashCode());
            Assert.AreNotEqual(object1.GetHashCode(), object3.GetHashCode());
        }

        [Test]
        public void TestToString()
        {
            UnitQuaternion test1 = new UnitQuaternion(1.0, 0.0, 0.0, 0.0);
            UnitQuaternion test2 = new UnitQuaternion(0.0, 1.0, 0.0, 0.0);
            UnitQuaternion test3 = new UnitQuaternion(0.0, 0.0, 1.0, 0.0);
            UnitQuaternion test4 = new UnitQuaternion(0.0, 0.0, 0.0, 1.0);

            Assert.AreEqual("1, 0, 0, 0", test1.ToString());
            Assert.AreEqual("0, 1, 0, 0", test2.ToString());
            Assert.AreEqual("0, 0, 1, 0", test3.ToString());
            Assert.AreEqual("0, 0, 0, 1", test4.ToString());
        }
    }
}

File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestUnitQuaternion.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.Action;
import agi.foundation.compatibility.annotations.CS2JWarning;
import agi.foundation.compatibility.AssertHelper;
import agi.foundation.compatibility.IEquatable;
import agi.foundation.compatibility.NotFiniteNumberException;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.TypeLiteral;
import cesiumlanguagewriter.*;
import javax.annotation.Nonnull;
import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.rules.ExpectedException;
import org.junit.runners.MethodSorters;
import org.junit.Test;

@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestUnitQuaternion {
    /**
    * Tests that initialization of and access to the type's values works correctly.
    */
    @Test
    public final void testHoldValue() {
        UnitQuaternion quaternion = new UnitQuaternion(2.0, 4.0, 7.0, 10.0);
        AssertHelper.assertEquals(2.0 / 13.0, quaternion.getW());
        AssertHelper.assertEquals(4.0 / 13.0, quaternion.getX());
        AssertHelper.assertEquals(7.0 / 13.0, quaternion.getY());
        AssertHelper.assertEquals(10.0 / 13.0, quaternion.getZ());
    }

    @Test
    public final void testInitializeAndReturnMagnitude() {
        double magnitude = 0D;
        final double[] out$magnitude$0 = new double[1];
        UnitQuaternion quaternion = new UnitQuaternion(2.0, 4.0, 7.0, 10.0, out$magnitude$0);
        magnitude = out$magnitude$0[0];
        AssertHelper.assertEquals(2.0 / 13.0, quaternion.getW(), Constants.Epsilon15);
        AssertHelper.assertEquals(4.0 / 13.0, quaternion.getX(), Constants.Epsilon15);
        AssertHelper.assertEquals(7.0 / 13.0, quaternion.getY(), Constants.Epsilon15);
        AssertHelper.assertEquals(10.0 / 13.0, quaternion.getZ(), Constants.Epsilon15);
        AssertHelper.assertEquals(13.0, magnitude, Constants.Epsilon15);
    }

    public final void testFromMatrix3By3(double angleDegrees, double axisX, double axisY, double axisZ) {
        Cartesian axis = new Cartesian(axisX, axisY, axisZ);
        UnitCartesian unit = axis.normalize();
        double angle = angleDegrees * Constants.RadiansPerDegree;
        double c = Math.cos(angle);
        double s = Math.sin(angle);
        double w = c;
        double x = s * unit.getX();
        double y = s * unit.getY();
        double z = s * unit.getZ();
        UnitQuaternion quaternion = new UnitQuaternion(w, x, y, z);
        Matrix3By3 matrix = new Matrix3By3(quaternion);
        UnitQuaternion test = new UnitQuaternion(matrix);
        AssertHelper.assertEquals(w, quaternion.getW(), Constants.Epsilon15);
        AssertHelper.assertEquals(x, quaternion.getX(), Constants.Epsilon15);
        AssertHelper.assertEquals(y, quaternion.getY(), Constants.Epsilon15);
        AssertHelper.assertEquals(z, quaternion.getZ(), Constants.Epsilon15);
    }

    @Test
    public final void testFromMatrix3By3$TestCase1() {
        testFromMatrix3By3(60.0, 2.0, 3.0, 6.0);
    }

    @Test
    public final void testFromMatrix3By3$TestCase2() {
        testFromMatrix3By3(120.0, 6.0, -3.0, -2.0);
    }

    @Test
    public final void testFromMatrix3By3$TestCase3() {
        testFromMatrix3By3(120.0, -2.0, -3.0, 6.0);
    }

    @Test
    public final void testFromMatrix3By3$TestCase4() {
        testFromMatrix3By3(120.0, -2.0, 6.0, -3.0);
    }

    /**
    * Tests the equality and inequality methods and operators.
    */
    @CS2JWarning("Unhandled attribute removed: SuppressMessage")
    @Test
    public final void testEquality() {
        UnitQuaternion first = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
        UnitQuaternion second = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
        AssertHelper.assertEquals(first, second);
        AssertHelper.assertEquals(second, first);
        Assert.assertTrue(UnitQuaternion.equals(first, second));
        Assert.assertTrue(UnitQuaternion.equals(second, first));
        Assert.assertFalse(UnitQuaternion.notEquals(first, second));
        Assert.assertFalse(UnitQuaternion.notEquals(second, first));
        Assert.assertTrue(first.equalsType(second));
        Assert.assertTrue(second.equalsType(first));
        second = new UnitQuaternion(1.1, 2.0, 3.0, 4.0);
        AssertHelper.assertNotEqual(first, second);
        AssertHelper.assertNotEqual(second, first);
        Assert.assertFalse(UnitQuaternion.equals(first, second));
        Assert.assertFalse(UnitQuaternion.equals(second, first));
        Assert.assertTrue(UnitQuaternion.notEquals(first, second));
        Assert.assertTrue(UnitQuaternion.notEquals(second, first));
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        second = new UnitQuaternion(1.0, 2.2, 3.0, 4.0);
        AssertHelper.assertNotEqual(first, second);
        AssertHelper.assertNotEqual(second, first);
        Assert.assertFalse(UnitQuaternion.equals(first, second));
        Assert.assertFalse(UnitQuaternion.equals(second, first));
        Assert.assertTrue(UnitQuaternion.notEquals(first, second));
        Assert.assertTrue(UnitQuaternion.notEquals(second, first));
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        second = new UnitQuaternion(1.0, 2.0, 3.3, 4.0);
        AssertHelper.assertNotEqual(first, second);
        AssertHelper.assertNotEqual(second, first);
        Assert.assertFalse(UnitQuaternion.equals(first, second));
        Assert.assertFalse(UnitQuaternion.equals(second, first));
        Assert.assertTrue(UnitQuaternion.notEquals(first, second));
        Assert.assertTrue(UnitQuaternion.notEquals(second, first));
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        second = new UnitQuaternion(1.0, 2.0, 3.3, 4.4);
        AssertHelper.assertNotEqual(first, second);
        AssertHelper.assertNotEqual(second, first);
        Assert.assertFalse(UnitQuaternion.equals(first, second));
        Assert.assertFalse(UnitQuaternion.equals(second, first));
        Assert.assertTrue(UnitQuaternion.notEquals(first, second));
        Assert.assertTrue(UnitQuaternion.notEquals(second, first));
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
    }

    @Test
    public final void testEqualsEpsilon() {
        UnitQuaternion first = new UnitQuaternion(1.0, 1.0, 1.0, 1.0);
        UnitQuaternion second = new UnitQuaternion(0.99, 1.0, 1.0, 1.01);
        Assert.assertTrue(second.equalsEpsilon(first, Constants.Epsilon1));
        Assert.assertTrue(second.equalsEpsilon(first, Constants.Epsilon2));
        Assert.assertFalse(second.equalsEpsilon(first, Constants.Epsilon3));
        Assert.assertFalse(second.equalsEpsilon(first, Constants.Epsilon4));
        Assert.assertFalse(second.equalsEpsilon(first, Constants.Epsilon5));
    }

    /**
    * Tests to ensure the equality fails when comparing incorrect type.
    */
    @CS2JWarning("Unhandled attribute removed: SuppressMessage")
    @Test
    public final void testEqualityWithWrongType() {
        UnitQuaternion first = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
        Cartographic second = new Cartographic(1.0, 2.0, 3.0);
        Assert.assertFalse(first.equals(second));
    }

    /**
    * Tests that rotation by an invalid value produces an
    {@link NotFiniteNumberException}.
    */
    @Test
    public final void testFromInfinity() {
        AssertHelper.<NotFiniteNumberException> assertThrows(new TypeLiteral<NotFiniteNumberException>() {}, Action.of(() -> {
            UnitQuaternion unused = new UnitQuaternion(Double.POSITIVE_INFINITY, 0.0, 0.0, 0.0);
        }));
    }

    @Test
    public final void testConjugate() {
        UnitQuaternion original = new UnitQuaternion(2.0, 4.0, 7.0, 10.0);
        UnitQuaternion conjugate = original.conjugate();
        AssertHelper.assertEquals(2.0 / 13.0, conjugate.getW());
        AssertHelper.assertEquals(-4.0 / 13.0, conjugate.getX());
        AssertHelper.assertEquals(-7.0 / 13.0, conjugate.getY());
        AssertHelper.assertEquals(-10.0 / 13.0, conjugate.getZ());
    }

    @Test
    public final void testIdentity() {
        UnitQuaternion identity = UnitQuaternion.getIdentity();
        AssertHelper.assertEquals(1.0, identity.getW());
        AssertHelper.assertEquals(0.0, identity.getX());
        AssertHelper.assertEquals(0.0, identity.getY());
        AssertHelper.assertEquals(0.0, identity.getZ());
    }

    @Test
    public final void testIsUndefined() {
        Assert.assertFalse(new UnitQuaternion(1.0, 1.0, 1.0, 1.0).getIsUndefined());
        Assert.assertTrue(UnitQuaternion.getUndefined().getIsUndefined());
        Assert.assertTrue(new UnitQuaternion(Double.NaN, 1.0, 1.0, 1.0).getIsUndefined());
        Assert.assertTrue(new UnitQuaternion(1.0, Double.NaN, 1.0, 1.0).getIsUndefined());
        Assert.assertTrue(new UnitQuaternion(1.0, 1.0, Double.NaN, 1.0).getIsUndefined());
        Assert.assertTrue(new UnitQuaternion(1.0, 1.0, 1.0, Double.NaN).getIsUndefined());
    }

    @Test
    public final void testNegation() {
        UnitQuaternion u = UnitQuaternion.negate(new UnitQuaternion(2.0, 4.0, 7.0, 10.0));
        AssertHelper.assertEquals(-2.0 / 13.0, u.getW());
        AssertHelper.assertEquals(-4.0 / 13.0, u.getX());
        AssertHelper.assertEquals(-7.0 / 13.0, u.getY());
        AssertHelper.assertEquals(-10.0 / 13.0, u.getZ());
    }

    @Test
    public final void testMultiplicationByUnitQuaternion() {
        // Choose quaternions whose vector portions (x, y, z) do not produce zeros
        // when dotted or crossed.  They should also have non-zero scalar (w) portions as well.
        UnitQuaternion first = new UnitQuaternion(7.0, 2.0, 3.0, 6.0);
        UnitQuaternion second = new UnitQuaternion(7.0, 2.0, 3.0, 6.0);
        UnitQuaternion result = first.multiply(second);
        AssertHelper.assertEquals(0.0 / 7.0, result.getW(), Constants.Epsilon15);
        AssertHelper.assertEquals(2.0 / 7.0, result.getX(), Constants.Epsilon15);
        AssertHelper.assertEquals(3.0 / 7.0, result.getY(), Constants.Epsilon15);
        AssertHelper.assertEquals(6.0 / 7.0, result.getZ(), Constants.Epsilon15);
        result = UnitQuaternion.multiply(first, second);
        AssertHelper.assertEquals(0.0 / 7.0, result.getW(), Constants.Epsilon15);
        AssertHelper.assertEquals(2.0 / 7.0, result.getX(), Constants.Epsilon15);
        AssertHelper.assertEquals(3.0 / 7.0, result.getY(), Constants.Epsilon15);
        AssertHelper.assertEquals(6.0 / 7.0, result.getZ(), Constants.Epsilon15);
    }

    /**
    * Tests that GetHashCode returns something at least reasonably random.
    */
    @Test
    public final void testGetHashCode() {
        UnitQuaternion object1 = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
        UnitQuaternion object2 = new UnitQuaternion(1.0, 2.0, 3.0, 4.0);
        UnitQuaternion object3 = new UnitQuaternion(1.0, 2.0, 3.0, 4.1);
        AssertHelper.assertEquals(object1.hashCode(), object2.hashCode());
        AssertHelper.assertNotEqual(object1.hashCode(), object3.hashCode());
    }

    @Test
    public final void testToString() {
        UnitQuaternion test1 = new UnitQuaternion(1.0, 0.0, 0.0, 0.0);
        UnitQuaternion test2 = new UnitQuaternion(0.0, 1.0, 0.0, 0.0);
        UnitQuaternion test3 = new UnitQuaternion(0.0, 0.0, 1.0, 0.0);
        UnitQuaternion test4 = new UnitQuaternion(0.0, 0.0, 0.0, 1.0);
        AssertHelper.assertEquals("1, 0, 0, 0", test1.toString());
        AssertHelper.assertEquals("0, 1, 0, 0", test2.toString());
        AssertHelper.assertEquals("0, 0, 1, 0", test3.toString());
        AssertHelper.assertEquals("0, 0, 0, 1", test4.toString());
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }

    @Nonnull
    private final ExpectedException rule$expectedException = ExpectedException.none();

    @Nonnull
    @Rule
    public ExpectedException getRule$expectedException() {
        return rule$expectedException;
    }
}